src/psyclone/domain/common/transformations/kernel_module_inline_trans.py

```diff
@@ -83,34 +85,75 @@ def validate(self, node, options=None):
         :param options: a dictionary with options for transformations.
         :type options: Optional[Dict[str, Any]]
 
-        :raises TransformationError: if the target node is not a sub-class of \
-            psyGen.CodedKern.
-        :raises TransformationError: if the subroutine containing the \
-            implementation of the kernel cannot be retrieved with \
-            'get_kernel_schedule'.
-        :raises TransformationError: if the name of the routine that \
-            implements the kernel is not the same as the kernel name. This \
-            will happen if the kernel is polymorphic (uses a Fortran \
+        :raises TransformationError: if the target node is not a sub-class of
+            psyGen.CodedKern or psyir.nodes.Call or is an IntrinsicCall.
+        :raises TransformationError: if the target node is not within a
+            Container (Fortran module).
+        :raises TransformationError: if there is no explicit import of the
+            called Routine and there is already a Routine of that name in the
+            parent Container.
+        :raises TransformationError: if the PSyIR of the implementation of the
+            called Routine/kernel cannot be retrieved.
+        :raises TransformationError: if the name of the routine that
+            implements the kernel is not the same as the kernel name. This
+            will happen if the kernel is polymorphic (uses a Fortran
             INTERFACE) and will be resolved by #1824.
         :raises TransformationError: if the kernel cannot be safely inlined.
 
         '''
-        if not isinstance(node, CodedKern):
+        if isinstance(node, CodedKern):
+            routine_sym = None
+            kname = node.name
+            kern_or_call = "Kernel"
+        elif isinstance(node, Call):
+            if isinstance(node, IntrinsicCall):
+                raise TransformationError(
+                    f"Cannot module-inline a call to an intrinsic (got "
+                    f"'{node.debug_string()}')")
+            routine_sym = node.routine
+            kname = routine_sym.name
+            kern_or_call = "routine"
+        else:
             raise TransformationError(
                 f"Target of a {self.name} must be a sub-class of "
-                f"psyGen.CodedKern but got '{type(node).__name__}'")
+                f"psyGen.CodedKern or psyir.nodes.Call but got "
+                f"'{type(node).__name__}'")
+
+        parent_container = node.ancestor(Container)
+        if isinstance(parent_container, FileContainer):
```

Comment:
Do we need `... or not parent_container)` in case it is a detached subtree?

Reply:
Oh yes, added.